File: ci-runners/codebuild_project.py
import pulumi
from pulumi import Input
from typing import Optional, Dict, TypedDict, Any
import pulumi_aws as aws

class CodebuildProjectArgs(TypedDict, total=False):
    name: Input[str]
    serviceRole: Input[str]
    computeType: Input[str]
    repository: Input[str]
    codeConnectionArn: Input[Any]

class CodebuildProject(pulumi.ComponentResource):
    def __init__(self, name: str, args: CodebuildProjectArgs, opts:Optional[pulumi.ResourceOptions] = None):
        super().__init__("components:index:CodebuildProject", name, args, opts)

        ubuntu_project = aws.codebuild.Project(f"{name}-ubuntu_project",
            name=args["name"],
            service_role=args["serviceRole"],
            artifacts={
                "type": "NO_ARTIFACTS",
            },
            cache={
                "type": "NO_CACHE",
            },
            build_timeout=60 * 6,
            environment={
                "compute_type": args["computeType"],
                "image": "aws/codebuild/standard:7.0-25.01.30",
                "type": "LINUX_CONTAINER",
                "image_pull_credentials_type": "CODEBUILD",
                "privileged_mode": True,
            },
            logs_config={
                "cloudwatch_logs": {
                    "status": "DISABLED",
                },
            },
            source={
                "type": "GITHUB",
                "location": f"https://github.com/{args["repository"]}",
                "git_clone_depth": 1,
                "git_submodules_config": {
                    "fetch_submodules": False,
                },
                "auth": {
                    "type": "CODECONNECTIONS",
                    "resource": args["codeConnectionArn"],
                },
            },
            opts = pulumi.ResourceOptions(parent=self))

        ubuntu_project_webhook = aws.codebuild.Webhook(f"{name}-ubuntu_project_webhook",
            project_name=ubuntu_project.name,
            build_type="BUILD",
            filter_groups=[{
                "filters": [{
                    "type": "EVENT",
                    "pattern": "WORKFLOW_JOB_QUEUED",
                }],
            }],
            opts = pulumi.ResourceOptions(parent=self))

        self.register_outputs()

File: ci-runners-ts/codebuild-project.ts
import * as pulumi from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";

export type CodebuildProjectArgs = {
    /**
     * The name of the project. What you need to write in the yaml GitHub Actions workflow.
     */
    name: pulumi.Input<string>;
    /**
     * The IAM role that AWS CodeBuild assumes to run the build.
     */
    serviceRole: pulumi.Input<string>;
    /**
     * Instance type.
     */
    computeType: pulumi.Input<string>;
    /**
     * The GitHub repository where the codebuild project will be used.
     */
    repository: pulumi.Input<string>;
    /**
     * Arn of the GitHub CodeConnection for the GitHub organization.
     */
    codeConnectionArn: pulumi.Input<any>;
}

export function createCodebuildProject(name: string, args: CodebuildProjectArgs, opts?: pulumi.ComponentResourceOptions): pulumi.ComponentResource {
    const component = new pulumi.ComponentResource("components:index:CodebuildProject", name, args, opts);

    const ubuntuProject = new aws.codebuild.Project(`${name}-ubuntu_project`, {
        name: args.name,
        serviceRole: args.serviceRole,
        artifacts: {
            type: "NO_ARTIFACTS",
        },
        cache: {
            type: "NO_CACHE",
        },
        buildTimeout: 60 * 6,
        environment: {
            computeType: args.computeType,
            image: "aws/codebuild/standard:7.0-25.01.30",
            type: "LINUX_CONTAINER",
            imagePullCredentialsType: "CODEBUILD",
            privilegedMode: true,
        },
        logsConfig: {
            cloudwatchLogs: {
                status: "DISABLED",
            },
        },
        source: {
            type: "GITHUB",
            location: `https://github.com/${args.repository}`,
            gitCloneDepth: 1,
            gitSubmodulesConfig: {
                fetchSubmodules: false,
            },
            auth: {
                type: "CODECONNECTIONS",
                resource: args.codeConnectionArn,
            },
        },
    }, {
        parent: component,
    });

    new aws.codebuild.Webhook(`${name}-ubuntu_project_webhook`, {
        projectName: ubuntuProject.name,
        buildType: "BUILD",
        filterGroups: [{
            filters: [{
                type: "EVENT",
                pattern: "WORKFLOW_JOB_QUEUED",
            }],
        }],
    }, {
        parent: component,
    });

    return component;
}
